pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -107,16 +110,39 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 		return reconcile.Result{}, err
 	}
 
-	var createdOrUpdated bool
-	var user *userv1.User
-	if user, createdOrUpdated, err = r.ensureUser(reqLogger, userAcc); err != nil || createdOrUpdated {
-		return reconcile.Result{}, err
-	}
+	// Check if the UserAccount has been deleted. If it has been deleted, delete
+	// secondary resources dentity and user.
+	if !userAcc.ObjectMeta.DeletionTimestamp.IsZero() {
+		var deleted bool
+		if err, deleted = r.deleteIdentity(reqLogger, userAcc); err != nil || deleted {
+			return reconcile.Result{}, err
+		}
 
-	if _, createdOrUpdated, err = r.ensureIdentity(reqLogger, userAcc, user); err != nil || createdOrUpdated {
-		return reconcile.Result{}, err
+		if err, deleted = r.deleteUser(reqLogger, userAcc); err != nil || deleted {
+			return reconcile.Result{}, err
+		}
+
+		if err = r.deleteFinalizer(reqLogger, userAcc); err != nil {
+			return reconcile.Result{}, err
+		}
 	}
 
+	if userAcc.ObjectMeta.DeletionTimestamp.IsZero() {
+		var createdOrUpdated bool
+		var user *userv1.User
+		if user, createdOrUpdated, err = r.ensureUser(reqLogger, userAcc); err != nil || createdOrUpdated {
+			return reconcile.Result{}, err
+		}
+
+		if _, createdOrUpdated, err = r.ensureIdentity(reqLogger, userAcc, user); err != nil || createdOrUpdated {
+			return reconcile.Result{}, err
+		}
+
+		// Add the finalizer if it is not present
+		if err := r.setFinalizers(userAcc); err != nil {
```

Comment:
call to `setFinalizers` should be first .. i.e. before `ensureUser` and `ensureIdentity`

Reply:
Fixed here: 008771ba402bc3dafb894c1eb8d0a7851f07df0d